builder: fix processing of invalid substitution syntax

thaJeztah opened the pull request:
fixes https://github.com/moby/moby/issues/37128

The builder did not detect syntax errors in substitutions in the
Dockerfile, causing those values to be processed incorrectly instead of
producing an error.

### Example 1: missing `}`

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa:-bbb
    RUN echo $var
    EOF

#### Before:

    Step 3/3 : RUN echo $var
     ---> Running in f06571e77146
    bbb

#### After:

    Step 2/3 : ARG var=${aaa:-bbb
    failed to process "${aaa:-bbb": syntax error: missing '}'

### Example 2: missing closing `}`, no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": syntax error: missing '}'

### Example 3: double opening bracket (`{`)

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa:-bbb}
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": syntax error: bad substitution

### Example 4: double opening bracket (`{`), no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa}
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": syntax error: bad substitution


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

```Markdown
* Fix parsing of invalid environment variable substitution 
```

thaJeztah committed:
builder: fix processing of invalid substitusion syntax

The builder did not detect syntax errors in substitusions in the
Dockerfile, causing those values to be processed incorrectly instead of
producing an error.

Example 1: missing `}`

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa:-bbb
    RUN echo $var
    EOF

Before:

    Step 3/3 : RUN echo $var
     ---> Running in f06571e77146
    bbb

After:

    Step 2/3 : ARG var=${aaa:-bbb
    failed to process "${aaa:-bbb": syntax error: missing '}'

Example 2: missing closing `}`, no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa
    RUN echo $var
    EOF

Before:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": missing ':' in substitution

After:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": syntax error: missing '}'

Example 3: double opening bracket (`{`)

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa:-bbb}
    RUN echo $var
    EOF

Before:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": missing ':' in substitution

After:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": syntax error: bad substitution

Example 4: double opening bracket (`{`), no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa}
    RUN echo $var
    EOF

Before:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": missing ':' in substitution

After:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": syntax error: bad substitution

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
ping @duglin PTAL

vdemeester reviewed (approved):
LGTM :tiger:

thaJeztah commented:
@rcjsuen  spotted one more omission in the validation; https://github.com/moby/moby/issues/37128#issuecomment-391734790 - let me add an extra commit to cover that as well

thaJeztah commented on builder/dockerfile/shell/envVarTest:
```diff
@@ -119,3 +119,22 @@ A|안녕${XXX:-\$PWD:}xx       |     안녕$PWD:xx
+A|${-:bbb}                  |     error
+A|${+:bbb}                  |     error
+A|${+}                      |     error
+A|${-}                      |     error
```
Should already be caught, but I just added (never hurts)

thaJeztah commented:
Working on some changes, after discussing this on slack.

ijc commented on builder/dockerfile/shell/lex.go:
```diff
@@ -261,6 +261,31 @@ func (sw *shellWord) processDollar() (string, error) {
+		//
+		// Note: validity of ${-:bbb} and ${-} is inconsistent between implementations.
+		// Treat it as invalid, because we cannot substitute the `-` with
+		// anything useful.
```
Why is `$-` special but `$*` `$@` `$#` `$?` `$$` `$0` and `$_` (other such special parameters in POSIX) do not appear to be?

I also think it is a mistake to conflate the existence/behaviour of those special parameters with the entirely unrelated support for the `${parameter:offset}` syntax, which is non-POSIX and hence not supported by dash but is supported by `bash` and whatever the alpine/busybox shell is these days (`ash`?).

Don't be confused into thinking that `${-:offset}` is somehow making `-:` into something special, it is not, it's just two unrelated concepts being used at the same time.

I think you should decide whether or not the `${parameter:offset}` form is valid, irrespective of whether `parameter` is `-` or `foo` or `PATH` and just implement what you chose there.

Then, completely separately, decide which, if any, special parameters are supported. This could be as simple as having no way to set `-` or `@` etc and therefore treating them as undefined variables at this level or could be done by seeing those names, whenever they are used, and rejecting them.

BTW I don't think `${:xx}` is valid anywhere, is it? Because it is missing the parameter name entirely so it is just syntactically invalid.

thaJeztah committed:
Fix detection for missing parameter in substitution

`${}`, `${:}` and so on are invalid because there's
no parameter within the brackets; fix detection for
this situation and add/update tests.

There were some existing test-cases that were testing
for the wrong behavior, which are now updated.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Add line-numbers to  asserts

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Add more test-cases for positional parameters

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Add detection of "special parameters" for substitution

Detect Special parameters as defined in
http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_05_02

Treat these as parameters that are not set, instead of
producing an error that a modifier is missing.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Handle multi-digit positional parameters

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@ijc updated PTAL; let me know what you think 👍 👎

tianon reviewed (approved):
That is a lot of new edge case tests! :heart_eyes:

thaJeztah commented:
They're unit-tests so didn't break the bank :joy:

thaJeztah commented:
ping @tonistiigi in case you need changes in BuildKit for this 👍

thaJeztah commented:
> buildkit uses the same package

d'oh! I looked for that last week, forgot I did 😂

vieux reviewed (approved):
LGTM @tonistiigi I'll let you merge since you own the builder.
